genrl/deep/common/base_class.py

```diff
@@ -0,0 +1,164 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Tuple, Union
+
+import numpy as np
+import torch
+
+from genrl.deep.common.utils import set_seeds
+
+
+class BaseAgent(ABC):
+    def __init__(
+        self,
+        network_type: str,
+        env: Any,
+        epochs: int = 100,
+        seed: Optional[int] = None,
+        render: bool = False,
+        device: Union[torch.device, str] = "cpu",
+        run_num: int = None,
+        save_model: str = None,
+        load_model: str = None,
+        save_interval: int = 50,
+    ):
+        self.network_type = network_type
+        self.env = env
+        self.epochs = epochs
+        self.seed = seed
+        self.render = render
+        self.run_num = run_num
+        self.save_model = save_model
+        self.load_model = load_model
+        self.save_interval = save_interval
+        self.observation_space = None
+        self.action_space = None
+
+        # Assign device
+        if "cuda" in device and torch.cuda.is_available():
+            self.device = torch.device(device)
+        else:
+            self.device = torch.device("cpu")
+
+        # Assign seed
+        if seed is not None:
+            set_seeds(seed, self.env)
+
+    def create_model(self) -> None:
+        """
+        Initialize all the policy networks in this method, and also \
+        initialize the optimizers and the buffers.
+        """
+        raise NotImplementedError()
+
+    def _update_params_before_select_action(self, timestep: int) -> None:
+        """
+        Update any parameters before selecting action like epsilon for decaying epsilon greedy
+
+        :param timestep: Timestep in the training process
+        :type timestep: int
+        """
+        pass
```

Comment:
Should raise a notimplementederror?

Reply:
No, because it won't be defined in every agent.